Add password confirmation in Get-Credential

Shriram0908 opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

## PR Summary

<!-- Summarize your PR between here and the checklist. -->
 ```Get-Credential``` : Add parameter ```-ReEnterPassword```  to resolve [#10625](https://github.com/PowerShell/PowerShell/issues/10625)

## PR Context

<!-- Provide a little reasoning as to why this Pull Request helps and why you have opened it. -->
By adding ```-ReEnterPassword``` parameter the user will be prompted to re-enter the password and performs a comparison. If comparison fails ```null``` is returned and user is notified about password mismatch.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: <!-- Number/link of that issue here --> [#6034](https://github.com/MicrosoftDocs/PowerShell-Docs/issues/6034)
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

Shriram0908 committed:
Get-Credential: Add parameter -ReEnterPassword to prompt user to re-enter the password for confirmation

iSazonov commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs:
```diff
@@ -25,6 +25,7 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
         /// if so configured.
         /// </summary>
         /// <param name="userName">Name of the user whose creds are to be prompted for. If set to null or empty string, the function will prompt for user name first.</param>
+        /// <param name="reEnterPassword">Promts user to re-enter the password for confirmation</param>
```
Typo.

iSazonov commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs:
```diff
@@ -48,6 +51,7 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
         /// Prompt for credentials.
         /// </summary>
         /// <param name="userName">Name of the user whose creds are to be prompted for. If set to null or empty string, the function will prompt for user name first.</param>
+        /// <param name="reEnterPassword">Promts user to re-enter the password for confirmation</param>
```
Typo.

iSazonov commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs:
```diff
@@ -112,7 +119,50 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
             {
                 return null;
             }
+            if (reEnterPassword)
```
Above we try to read user name in a cycle until we get valid name or Crtl-C (null).
I'd expect the same behavior for password too.

iSazonov commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs:
```diff
@@ -112,7 +119,50 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
+            {
+                reEnterPasswordPrompt = StringUtil.Format(ConsoleHostUserInterfaceSecurityResources.PromptForCredential_ReEnterPassword, userName);
+
+                //
+                // now, prompt to re-enter the password
+                //
```
Please use one line comments like:
```suggestion
                // Now, prompt to re-enter the password.
```

iSazonov commented on src/Microsoft.PowerShell.Security/security/CredentialCommands.cs:
```diff
@@ -79,6 +79,19 @@ public string Title
+        /// Gets and sets the Re-enter Password on the window promt.
+        /// </summary>
+        [Parameter(Mandatory = false, ParameterSetName = messageSet)]
+        public SwitchParameter ReEnterPassword
+        {
+            get { return _reEnterPassword; }
+
+            set { _reEnterPassword = value; }
+        }
+
+        private bool _reEnterPassword;
```
Please use auto-implemented property:
```suggestion
        public SwitchParameter ReEnterPassword { get; set; }
```

iSazonov commented on src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs:
```diff
@@ -830,9 +833,9 @@ private void FlushPendingOutput()
-        /// <seealso cref="System.Management.Automation.Host.PSHostUserInterface.PromptForCredential(string, string, string, string, System.Management.Automation.PSCredentialTypes, System.Management.Automation.PSCredentialUIOptions)"/>
+        /// <seealso cref="System.Management.Automation.Host.PSHostUserInterface.PromptForCredential(string, string, string, bool, string, System.Management.Automation.PSCredentialTypes, System.Management.Automation.PSCredentialUIOptions)"/>
         public abstract PSCredential PromptForCredential(string caption, string message,
-            string userName, string targetName
+            string userName, bool reEnterPassword, string targetName
         );
```
It is a public API. We can not change public API - it is a breaking change. We can and new overload.

iSazonov commented on src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs:
```diff
@@ -848,6 +851,9 @@ private void FlushPendingOutput()
         /// string, the function will prompt for user name first.
         /// </param>
+        /// <param name="reEnterPassword">
+        /// Promts user to re-enter the password for confirmation
```
Please add final dot in comment:
```suggestion
        /// Prompts user to re-enter the password for confirmation.
```

iSazonov commented on src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs:
```diff
@@ -864,9 +870,9 @@ private void FlushPendingOutput()
-        /// <seealso cref="System.Management.Automation.Host.PSHostUserInterface.PromptForCredential(string, string, string, string)"/>
+        /// <seealso cref="System.Management.Automation.Host.PSHostUserInterface.PromptForCredential(string, string, string, bool, string)"/>
         public abstract PSCredential PromptForCredential(string caption, string message,
-            string userName, string targetName, PSCredentialTypes allowedCredentialTypes,
+            string userName, bool reEnterPassword, string targetName, PSCredentialTypes allowedCredentialTypes,
```
Please revert and add new overload.

iSazonov commented on src/System.Management.Automation/engine/hostifaces/internalHostuserInterfacesecurity.cs:
```diff
@@ -22,10 +22,11 @@ public override
             string caption,
             string message,
             string userName,
+            bool reEnterPassword,
```
The same.

iSazonov commented on src/System.Management.Automation/engine/hostifaces/internalHostuserInterfacesecurity.cs:
```diff
@@ -42,6 +43,7 @@ public override
             string caption,
             string message,
             string userName,
+            bool reEnterPassword,
```
The same.

iSazonov commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs:
```diff
@@ -59,14 +63,17 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
             string caption,
             string message,
             string userName,
+            bool reEnterPassword,
```
```suggestion
            bool reenterPassword,
```

iSazonov commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs:
```diff
@@ -59,14 +63,17 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
+            SecureString confirmPassword = null;
             string userPrompt = null;
             string passwordPrompt = null;
+            string reEnterPasswordPrompt = null;
```
```suggestion
            string reenterPasswordPrompt = null;
```

Shriram0908 committed:
Add new overload function PromptForCredential

 New overload for PromptForCredential in PSHostUserInterface class to avoid breaking change
 Add loops to ensure empty password is not provided
 Fix the typo and renamed reEnterPassword to reenterPassword

Shriram0908 commented:
@iSazonov Thank you for the feedback. I have made the appropriate changes. Please review.

I see the below error message in Linux test. I can see the file is present. What could be casing this issue.

```Pester v4.10.1
Executing all tests in 'test/powershell/Language/Operators/NullConditional.Tests.ps1', 'test/powershell/Language/Parser/Parsing.Tests.ps1', '/test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1' with Tags RequireSudoOnUnix
Resolve-Path: /home/vsts/work/1/a/bins/publish/Modules/Pester/4.10.1/Pester.psm1:1326
Line |
1326 |  …             & $script:SafeCommands['Resolve-Path'] -Path $unresolvedP …
     |                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot find path
     | '/test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1'
     | because it does not exist.



Tests completed in 191ms
Tests Passed: 0, Failed: 0, Skipped: 0, Pending: 0, Inconclusive: 0 
WARNING: Tests failed. See the issue below.
Exception: /home/vsts/work/1/s/tools/ci.psm1:672
Line |
 672 |  … ll -ne $_} | ForEach-Object { Test-PSPesterResults -ResultObject $_ }
     |                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Parameter cannot be processed because the parameter name
     | 'Error' is ambiguous. Possible matches include: -ErrorAction
     | -ErrorVariable.


##[error]PowerShell exited with code '1'.
```

iSazonov commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs:
```diff
@@ -103,16 +131,70 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
+
+                    // If the string is not equal print error message and returns null.
+
+                    if (!(pwd.Equals(confirmPwd)))
+                    {
+                        reenterPasswordPrompt = StringUtil.Format(ConsoleHostUserInterfaceSecurityResources.PasswordMismatch);
+                        WriteToConsole(reenterPasswordPrompt, true);
+                        return null;
+                    }
```
I'd expect a cycle here too - if user enter wrong confirmation we should prompt again for a password.

Shriram0908 replied:
@iSazonov  I have made the changes. Please review
``` 
Get-Credential -UserName user -ReEnterPassword

PowerShell credential request
Enter your credentials.
Password for user user: **
Re-Enter Password for user user:
Passwords do not match. Re-Enter Password for user user:
Passwords do not match. Re-Enter Password for user user:
Passwords do not match. Re-Enter Password for user user:
Passwords do not match. Re-Enter Password for user user:**
Passwords do not match. Re-Enter Password for user user:
Passwords do not match. Re-Enter Password for user user:
Passwords do not match. Re-Enter Password for user user:**


UserName                     Password
--------                     --------
user     System.Security.SecureString

Shriram0908 committed:
Get-Credential: Prompt user to re-enter password if mismatch while using -ReEnterPassword

The user will be prompted to re-enter the password if the password does not match.
Fix few CodeFactor issues

iSazonov commented on src/Microsoft.PowerShell.Security/security/CredentialCommands.cs:
```diff
@@ -100,7 +106,9 @@ protected override void BeginProcessing()
             try
             {
-                Credential = this.Host.UI.PromptForCredential(_title, _message, _userName, string.Empty);
+                Credential = this.Host.UI.PromptForCredential(_title, _message, _userName, ReEnterPassword, string.Empty,
+                    PSCredentialTypes.Default,
+                    PSCredentialUIOptions.Default);
```
```suggestion
                Credential = this.Host.UI.PromptForCredential(
                    _title,
                    _message,
                    _userName,
                    ReenterPassword,
                    string.Empty,
                    PSCredentialTypes.Default,
                    PSCredentialUIOptions.Default);
```

iSazonov commented on src/Microsoft.PowerShell.Security/security/CredentialCommands.cs:
```diff
@@ -79,6 +79,12 @@ public string Title
+        /// Gets or sets the Re-enter Password on the window prompt.
+        /// </summary>
+        [Parameter(Mandatory = false, ParameterSetName = messageSet)]
+        public SwitchParameter ReEnterPassword { get; set; }
```
```suggestion
        public SwitchParameter ReenterPassword { get; set; }
```

iSazonov commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:
```diff
@@ -87,6 +87,22 @@ public override PSCredential PromptForCredential(string caption, string message,
         }
 
+        /// <summary>
+        /// PromptForCredential.
```
```suggestion
        /// Prompt for credential.
```

iSazonov commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs:
```diff
@@ -100,19 +127,68 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
+                        // If the string is not equal print message and loop.
+                        if(!isPasswordMatch)
+                        {
+                            WriteToConsole(passwordMismatch, true);
```
Maybe use a "red" color?

Shriram0908 replied:
@iSazonov  I have added red color for "Passwords do not match." Since it was an error message I moved it to a separate line. Please review and let me know your feedback.
```
Get-Credential -UserName hello -ReenterPassword

PowerShell credential request
Enter your credentials.
Password for user hello: **
Re-Enter Password for user hello:
Passwords do not match.
Re-Enter Password for user hello:
Passwords do not match.
Re-Enter Password for user hello:**
Passwords do not match.
Re-Enter Password for user hello:**

iSazonov commented on src/Microsoft.PowerShell.ConsoleHost/resources/ConsoleHostUserInterfaceSecurityResources.resx:
```diff
@@ -123,4 +123,10 @@
+    <value>Re-Enter Password for user {0}:</value>
+  </data>
+  <data name="PromptForCredential_PasswordMismatch" xml:space="preserve">
+    <value>Passwords do not match. </value>
```
```suggestion
    <value>Passwords do not match.</value>
```

iSazonov commented on src/Microsoft.PowerShell.Security/security/CredentialCommands.cs:
```diff
@@ -79,6 +79,12 @@ public string Title
         private string _title = UtilsStrings.PromptForCredential_DefaultCaption;
 
+        /// <summary>
+        /// Gets or sets the Re-enter Password on the window prompt.
```
```suggestion
        /// Gets or sets the re-enter password on the window prompt.
```

Shriram0908 committed:
Improve code readability

Add new private function in ConsoleHostUserInterface to compare secure stings
Print "Passwords do not match." in red
Fix CodeFactor issues

iSazonov commented on test/tools/Modules/HelpersHostCS/HelpersHostCS.psm1:
```diff
@@ -145,18 +146,63 @@ namespace TestHost
+            SecureString password = ReadLineAsSecureString();
+            if(reenterPassword)
+            {
+                SecureString confirmPassword = ReadLineAsSecureString();
```
The test host emulates user IO for the cmdlet. That allows to test the cmdlet but not real host - real host can be tested only manually.

We need to check three scenario:
1. ReenterPassword is false
2. ReenterPassword is true and first password confirmation is right
3. ReenterPassword is true,  first password confirmation is wrong,  second password confirmation is right.

As I said, real host can be tested only manually so the test host can address only:
1. ReenterPassword is false
Test output is `Streams.Prompt.Add("Credential:" + caption + ":" + message);`
2. ReenterPassword is true
Test output could be `Streams.Prompt.Add("Credential@" + caption + "@" + message);`

Shriram0908 replied:
@iSazonov I have made changes to the test case and helper module. Please review and provide your feedback.

Shriram0908 committed:
Get-Credential: Add Pester test case

iSazonov commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs:
```diff
@@ -100,25 +125,76 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
+                    }
 
-            cred = new PSCredential(userName, password);
+                    if(SecureStringEquals(password, confirmPassword))
```
```suggestion
                    if (SecureStringEquals(password, confirmPassword))
```

Shriram0908 replied:
Sure. I will fix it. Anything else that I need to work on ?

iSazonov reviewed (commented):
LGTM with one minor comment.

iSazonov commented:
@Shriram0908 Please open new issue in PowerSehll-Docs repo and add the reference to the PR description.

Shriram0908 committed:
Fix code formating

PaulHigin commented:
Re-entering a password is normally done when creating a username/password, but not when using a password to access resources.  I am not sure how useful this change is, particularly since it is opt-in via a parameter switch.  I would like to get the committee's opinion before accepting the change.

PaulHigin commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs:
```diff
@@ -100,25 +125,76 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
+            try
+            {
+                pwd_ptr = Marshal.SecureStringToBSTR(password);
+                string pwd = Marshal.PtrToStringBSTR(pwd_ptr);
```
Converting passwords to plain text, leaves multiple immutable string objects in memory, indefinitely.  I think it would be better to convert to byte array for comparison, and the byte array values can be zeroed out afterwards.  Also you can compare byte array length for early out.

Shriram0908 replied:
@PaulHigin Thank you for the review. I have updated the function to use byte array for comparison and have zeroed it in the finally clause. The memory space where the pointer was pointing had a null value between each byte so I had to double the length to use as size. Please review and provide your feedback.

PaulHigin commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs:
```diff
@@ -100,25 +125,76 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
+            {
+                reenterPasswordPrompt = StringUtil.Format(ConsoleHostUserInterfaceSecurityResources.PromptForCredential_ReenterPassword, userName);
+                passwordMismatch = StringUtil.Format(ConsoleHostUserInterfaceSecurityResources.PromptForCredential_PasswordMismatch);
+                do
```
I wonder if the user should be prompting multiple times for re-entering password in this loop.  It is not known if the password was incorrect the first or second time it was entered.  So it may be better to exit with error immediately if the confirmation fails, and let the user try again from the beginning.

Shriram0908 replied:
I have changed the behavior.

iSazonov replied:
> and let the user try again from the beginning.

What is beginning? Go to user or password re-enter?

iSazonov commented:
@PaulHigin Have you any concerns about remoting scenarios?

PaulHigin commented:
@iSazonov  Good question.  I think this should work fine over remoting since the system will make remote host callbacks for prompting as before, but it would be good to check.

Shriram0908 committed:
Remove confirm password loop

the cmdlet return null if the confirm password does not match
uses byte array for comparison of password

iSazonov commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs:
```diff
@@ -170,17 +167,25 @@ public override PSCredential PromptForCredential(string caption, string message,
             {
                 pwd_ptr = Marshal.SecureStringToBSTR(password);
-                string pwd = Marshal.PtrToStringBSTR(pwd_ptr);
+                Marshal.Copy(pwd_ptr, pwd, 0, length);
```
We have no need to make the copy. See https://www.sjoerdlangkemper.nl/2017/11/08/comparing-securestrings-in-dotnet/
```c#
static bool IsEqual(IntPtr bstr1, IntPtr bstr2)
{
    var length1 = Marshal.ReadInt32(bstr1, -4);
    var length2 = Marshal.ReadInt32(bstr2, -4);

    if (length1 != length2) return false;

    var equal = 0;
    for (var i = 0; i < length1; i++)
    {
        var c1 = Marshal.ReadByte(bstr1 + i);
        var c2 = Marshal.ReadByte(bstr2 + i);
        equal |= c1 ^ c2;
    }
    return equal == 0;
}
```

Shriram0908 replied:
@iSazonov I have made the changes. Please review and let me know your feedback.

Shriram0908 committed:
Use bit comparison for secure sting

Remove byte array comparison method

iSazonov commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs:
```diff
@@ -167,25 +167,30 @@ public override PSCredential PromptForCredential(string caption, string message,
+                }
+
+                int equal = 0;
+                for(int i =0; i< pwdLength; i++)
```
```suggestion
                for(int i = 0; i < pwdLength; i++)
```

Shriram0908 committed:
Fix code formating

Co-authored-by: Ilya <darpa@yandex.ru>

PaulHigin commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs:
```diff
@@ -100,25 +126,84 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
+                    return false;
+                }
+
+                int equal = 0;
```
It is not clear to me that doing multiple PInvoke reads is that performant, or that we really need to worry about performance at this level.  For readability and maintainability I am inclined to simply copy the byte array and compare bytes (zero out array in finally block).  Also note that this loop can be exited as soon as one byte mismatch is found.

Shriram0908 replied:
@PaulHigin Is the below snippet as per requirements. 
```
            int size = password.Length * 2;
            IntPtr pwd_ptr = IntPtr.Zero;
            IntPtr confirmPwd_ptr = IntPtr.Zero;
            byte[] pwd = new byte[size];
            byte[] confirmPwd = new byte[size];
            try
            {
                pwd_ptr = Marshal.SecureStringToBSTR(password);
                Marshal.Copy(pwd_ptr, pwd, 0, size);

                confirmPwd_ptr = Marshal.SecureStringToBSTR(confirmPassword);
                Marshal.Copy(confirmPwd_ptr, confirmPwd, 0, size);

                return pwd.SequenceEqual(confirmPwd);
            }
            finally
            {
                if (pwd_ptr != IntPtr.Zero)
                {
                    Marshal.ZeroFreeBSTR(pwd_ptr);
                }

                if (confirmPwd_ptr != IntPtr.Zero)
                {
                    Marshal.ZeroFreeBSTR(confirmPwd_ptr);
                }

                Array.Clear(pwd, 0, size);
                Array.Clear(confirmPwd, 0, size);
                size = 0;
            }
```

iSazonov replied:
The pattern with `Marshal.ReadInt32` comes from StackOverflow site and could is used broadly. It and look good for me.

PaulHigin replied:
I apologize for the churn.  There are various ways to do this and I feel that perf differences are not significant.  I am fine with it either way and will leave it up to the author, however I do want to see an early out for first mismatch in the loop (which was missing before).

Shriram0908 replied:
@iSazonov @PaulHigin 
I have made the changes. Please review and let me know your feedback.

Shriram0908 committed:
Add early return if comparison fails

Shriram0908 committed:
Fix code formating

SteveL-MSFT commented:
@PowerShell/powershell-committee reviewed this, we agree on adding this capability, however we suggest that the switch should be called `-ConfirmPassword`

PaulHigin commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs:
```diff
@@ -100,25 +126,84 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
+            IntPtr confirmPwd_ptr = IntPtr.Zero;
+            try
+            {
+                pwd_ptr = Marshal.SecureStringToBSTR(password);
```
Overall looks good to me.  However, Marshal.SecureStringToBSTR() can return a IntPtr.Zero value.  I think we should check for that and return false if either pointer is that value.

PaulHigin commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs:
```diff
@@ -100,25 +126,84 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
+                int equal = 0;
+                for (int i = 0; i < pwdLength; i++)
+                {
+                    byte c1 = Marshal.ReadByte(pwd_ptr + i);
```
Marshal.ReadByte(), can also take an offset parameter, instead of incrementing the pointer reference.  Not sure if it makes a difference.
```
Marshal.ReadByte(pwd_ptr, i);
```

iSazonov replied:
Looks more readable.

Shriram0908 committed:
Rename from ReenterPassword to ConfirmPassword

iSazonov commented:
@PaulHigin Could you approve the PR?

iSazonov commented:
@Shriram0908 Please rebase to pass CI-static.

Shriram0908 committed:
Fix code factor issue

Shriram0908 commented:
> 
> 
> @Shriram0908 Please rebase to pass CI-static.

@iSazonov I rebased the branch but now mac os tests are failing. I don't think the failure has anything to do with my changes. I checked in Azure DevOps pipeline the error is related to test-connection and update-help

iSazonov commented:
@Shriram0908 CI-MacOS failure is know issue.

PaulHigin reviewed (approved):
LGTM

iSazonov commented:
@Shriram0908 Please rebase to pass CIs.

Shriram0908 commented:
> 
> 
> @Shriram0908 Please rebase to pass CIs.

@iSazonov Done... :)

iSazonov commented:
@Shriram0908 Thanks for your contribution!

Shriram0908 commented:
Thank you @iSazonov and @PaulHigin for your feedback.

rjmholt commented on src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs:
```diff
@@ -893,10 +900,57 @@ private void FlushPendingOutput()
+        /// <seealso cref="System.Management.Automation.Host.PSHostUserInterface.PromptForChoice"/>
+        /// <seealso cref="System.Management.Automation.Host.PSHostUserInterface.PromptForCredential(string, string, string, string)"/>
+        /// <seealso cref="System.Management.Automation.Host.PSHostUserInterface.PromptForCredential(string, string, string, string, System.Management.Automation.PSCredentialTypes, System.Management.Automation.PSCredentialUIOptions)"/>
+        public abstract PSCredential PromptForCredential(
+            string caption,
+            string message,
+            string userName,
+            bool confirmPassword,
+            string targetName,
+            PSCredentialTypes allowedCredentialTypes,
+            PSCredentialUIOptions options);
```
@TylerLeonhardt this may be why the PowerShell daily build has been causing CI issues.

There's a new abstract method in the host which has broken me:

```
An error has occurred that was not properly handled. Additional information is shown below. The PowerShell process will exit.
Unhandled exception. System.TypeLoadException: Method 'PromptForCredential' in type 'Microsoft.PowerShell.EditorServices.Services.PowerShell.Host.EditorServicesConsolePSHostUserInterface' from assembly 'Microsoft.PowerShell.EditorServices, Version=2.2.0.0, Culture=neutral, PublicKeyToken=null' does not have an implementation.
   at Microsoft.PowerShell.EditorServices.Services.PowerShell.Host.EditorServicesConsolePSHost..ctor(ILoggerFactory loggerFactory, String name, Version version, PSHost internalHost, ConsoleReadLine readline)
   at Microsoft.PowerShell.EditorServices.Services.PowerShell.PowerShellExecutionService.Initialize() in C:\Users\Robert Holt\Documents\Dev\Microsoft\PowerShellEditorServices\src\PowerShellEditorServices\Services\PowerShell\PowerShellExecutionService.cs:line 286
   at Microsoft.PowerShell.EditorServices.Services.PowerShell.PowerShellExecutionService.RunTopLevelConsumerLoop() in C:\Users\Robert Holt\Documents\Dev\Microsoft\PowerShellEditorServices\src\PowerShellEditorServices\Services\PowerShell\PowerShellExecutionService.cs:line 244
   at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart()
```

I'll open an issue to track fixing this

iSazonov replied:
Can we add new CI for ESS in the repo?

Shriram0908 replied:
@rjmholt , @iSazonov  Anything I could do to help ?

iSazonov replied:
@Shriram0908 Thanks! It is a question for MSFT team - they could add optional CIs for dependent tools so we could see possible regressions early.
